File: src/gameState.js
const { cloneBoard, placePiece, moveTopPiece } = require('./board');
const { pieceDefinitions, defaultReserve } = require('./pieces');
const { key } = require('./coords');

const createEmptyReserve = () => ({ ...defaultReserve });

const createGame = () => ({
  board: new Map(),
  reserves: {
    white: createEmptyReserve(),
    black: createEmptyReserve(),
  },
  turn: 0,
  currentPlayer: 'white',
  turnCounts: { white: 0, black: 0 },
  piecePositions: new Map(),
  moveHistory: [],
  pieceCounters: { white: 0, black: 0 },
});

const nextPlayer = (player) => (player === 'white' ? 'black' : 'white');

const applyPlacement = (state, player, pieceType, coord, pieceId) => {
  const next = { ...state, board: cloneBoard(state.board) };
  const piece = {
    id: pieceId,
    player,
    type: pieceType,
  };
  placePiece(next.board, coord, piece);
  next.reserves = {
    ...state.reserves,
    [player]: { ...state.reserves[player], [pieceType]: state.reserves[player][pieceType] - 1 },
  };
  next.piecePositions = new Map(state.piecePositions);
  next.piecePositions.set(pieceId, { coord });
  next.moveHistory = [...state.moveHistory, { type: 'place', player, pieceType, coord }];
  next.turnCounts = { ...state.turnCounts, [player]: state.turnCounts[player] + 1 };
  next.turn += 1;
  next.currentPlayer = nextPlayer(player);
  next.pieceCounters = { ...state.pieceCounters, [player]: state.pieceCounters[player] + 1 };
  return next;
};

const applyMovement = (state, player, pieceId, from, to) => {
  const next = { ...state, board: cloneBoard(state.board) };
  moveTopPiece(next.board, from, to);
  next.piecePositions = new Map(state.piecePositions);
  next.piecePositions.set(pieceId, { coord: to });
  next.moveHistory = [...state.moveHistory, { type: 'move', player, pieceId, from, to }];
  next.turnCounts = { ...state.turnCounts, [player]: state.turnCounts[player] + 1 };
  next.turn += 1;
  next.currentPlayer = nextPlayer(player);
  next.reserves = state.reserves;
  next.pieceCounters = state.pieceCounters;
  return next;
};

const pieceIsPlaced = (state, pieceId) => state.piecePositions.has(pieceId);

const playerHasPlacedQueen = (state, player) => {
  for (const stack of state.board.values()) {
    if (stack.some((p) => p.player === player && p.type === 'queen')) return true;
  }
  return false;
};

const nextPieceId = (state, player, pieceType) => {
  const count = state.pieceCounters[player] + 1;
  return `${player}-${pieceType}-${count}`;
};

module.exports = {
  createGame,
  applyPlacement,
  applyMovement,
  nextPlayer,
  pieceIsPlaced,
  playerHasPlacedQueen,
  nextPieceId,
};

File: src/game_state.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional

from .board import clone_board, move_top_piece, place_piece
from .coords import Coord
from .pieces import default_reserve


@dataclass
class GameState:
    board: Dict[str, List[dict]] = field(default_factory=dict)
    reserves: Dict[str, Dict[str, int]] = field(
        default_factory=lambda: {"white": dict(default_reserve), "black": dict(default_reserve)}
    )
    turn: int = 0
    current_player: str = "white"
    turn_counts: Dict[str, int] = field(default_factory=lambda: {"white": 0, "black": 0})
    piece_positions: Dict[str, dict] = field(default_factory=dict)
    move_history: List[dict] = field(default_factory=list)
    piece_counters: Dict[str, int] = field(default_factory=lambda: {"white": 0, "black": 0})


def create_game() -> GameState:
    return GameState()


def _clone_reserves(reserves: Dict[str, Dict[str, int]]) -> Dict[str, Dict[str, int]]:
    return {player: dict(pool) for player, pool in reserves.items()}


def clone_state(state: GameState, **overrides) -> GameState:
    return GameState(
        board=clone_board(state.board),
        reserves=_clone_reserves(state.reserves),
        turn=overrides.get("turn", state.turn),
        current_player=overrides.get("current_player", state.current_player),
        turn_counts=dict(state.turn_counts),
        piece_positions=dict(state.piece_positions),
        move_history=list(state.move_history),
        piece_counters=dict(state.piece_counters),
    )


def next_player(player: str) -> str:
    return "black" if player == "white" else "white"


def next_piece_id(state: GameState, player: str, piece_type: str) -> str:
    count = state.piece_counters[player] + 1
    return f"{player}-{piece_type}-{count}"


def apply_placement(state: GameState, player: str, piece_type: str, coord: Coord, piece_id: Optional[str] = None) -> GameState:
    next_state = GameState(
        board=clone_board(state.board),
        reserves=_clone_reserves(state.reserves),
        turn=state.turn,
        current_player=state.current_player,
        turn_counts=dict(state.turn_counts),
        piece_positions=dict(state.piece_positions),
        move_history=list(state.move_history),
        piece_counters=dict(state.piece_counters),
    )
    pid = piece_id or next_piece_id(state, player, piece_type)
    piece = {"id": pid, "player": player, "type": piece_type}
    place_piece(next_state.board, coord, piece)
    next_state.reserves[player][piece_type] -= 1
    next_state.piece_positions[pid] = {"coord": coord}
    next_state.move_history.append({"type": "place", "player": player, "pieceType": piece_type, "coord": coord})
    next_state.turn_counts[player] += 1
    next_state.turn += 1
    next_state.current_player = next_player(player)
    next_state.piece_counters[player] += 1
    return next_state


def apply_movement(state: GameState, player: str, piece_id: str, from_coord: Coord, to_coord: Coord) -> GameState:
    next_state = GameState(
        board=clone_board(state.board),
        reserves=_clone_reserves(state.reserves),
        turn=state.turn,
        current_player=state.current_player,
        turn_counts=dict(state.turn_counts),
        piece_positions=dict(state.piece_positions),
        move_history=list(state.move_history),
        piece_counters=dict(state.piece_counters),
    )
    move_top_piece(next_state.board, from_coord, to_coord)
    next_state.piece_positions[piece_id] = {"coord": to_coord}
    next_state.move_history.append({"type": "move", "player": player, "pieceId": piece_id, "from": from_coord, "to": to_coord})
    next_state.turn_counts[player] += 1
    next_state.turn += 1
    next_state.current_player = next_player(player)
    return next_state
